### PR TITLE
`path` function

### PR DESCRIPTION
**Pull request by adamjanicki2**
Addresses #38

### CODE REVIEW

File: src/path.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { properties } from "./children.js";
+
+/**
+ * Find a path from an ancestor node to a descendant node.
+ * @param {object} ancestor the ancestor node to start the search from
+ * @param {object} descendant the descendant node to search for
+ * @returns {(string | number)[] | null} an array of keys to traverse to get from the ancestor to the descendant, or null if no path exists
```
Comment (adamjanicki2) [inline src/path.js:7]: Should it return null or undefined if a path does not exist?

Comment (LeaVerou) [reply src/path.js:7]: `null`

File: src/path.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ */
+export default function path (ancestor, descendant) {
+	// Found descendant
+	if (ancestor === descendant) {
+		return [];
+	}
+	// Case where ancestor is an array for an array expression or compound etc
```
Comment (LeaVerou) [inline src/path.js:14]:
```suggestion

	// Case where ancestor is an array for an array expression or compound etc
```

File: src/path.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	}
+
+	const childProperties = properties[ancestor.type] ?? [];
+
+	for (const property of childProperties) {
+		const childPath = path(ancestor[property], descendant);
+		if (childPath) {
+			return [property].concat(childPath);
+		}
+	}
```
Comment (LeaVerou) [inline src/path.js:33]: This seems like two blocks of code doing largely the same thing.
Can we unify them? Array indices are properties after all. We could trivially unify them by using `childProperties = Array(ancestor.length).fill(0).map((_, i) => i);` for arrays, but that's a bit clumsy and a bit wasteful.

**Review by LeaVerou — CHANGES_REQUESTED** [review]
At first glance (keep in mind I'm reviewing this during a meeting) this seems a little wasteful: it's exploring more paths than are actually needed. There may be no other way if we don't have parent pointers, but if we do, we can short-circuit this process quite a lot.

**Comment by adamjanicki2** [comment]
> At first glance (keep in mind I'm reviewing this during a meeting) this seems a little wasteful: it's exploring more paths than are actually needed. There may be no other way if we don't have parent pointers, but if we do, we can short-circuit this process quite a lot.

Makes sense, I implemented like this to decouple it from the parents module, but I'll make a this function check to see if the descendant has parent pointers, and traverse them if so

File: test/path.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			args: [ast.right, ast.right.arguments[0]],
+			expect: ["arguments", 0],
+			description: "Non-root to leaf"
+		},
+		{
+			args: [ast, jsep("dne")],
+			expect: null,
+			description: "Non-existent path"
+		}
+	]
```
Comment (LeaVerou) [inline test/path.js:42]: Do we actually want the first argument here to be the name? Shouldn't what you have as `description` basically be the name?

Comment (adamjanicki2) [reply test/path.js:42]: Sure I can fix them

File: src/path.js (hunk 3 above)
Comment (adamjanicki2) [reply src/path.js:33]: I can't think of any other way to unify them without doing the trick you mentioned; I tried to think of one while implementing it, but couldn't think of one that wouldn't waste some performance

**Comment by adamjanicki2** [comment]
> At first glance (keep in mind I'm reviewing this during a meeting) this seems a little wasteful: it's exploring more paths than are actually needed. There may be no other way if we don't have parent pointers, but if we do, we can short-circuit this process quite a lot.

Something I didn't think about is the fact that if we try to use parents to our advantage, it's still messy because we'd have to examine all of the properties of a node to determine which one led to its child. So it would still probably save work on average since it's still walking up a single branch of the tree instead of walking down all of them, but I'm not sure it's worth the extra 20 lines of code, what do you think?

I think we should implement this without using parent pointers, and then come back to this once we've discussed #34 further. Augmenting parent pointers to contain a `property` and `index` field would make this doable!

**Comment by adamjanicki2** [comment]
Just pushed what it would look like if we have a version that checks for parent pointers; I wrote it pretty fast so it might be able to be trimmed down, but it's pretty lengthy

**Comment by LeaVerou** [comment]
I don’t have a ton of time to review right now, but at first glance that seems …excessive. Maybe we should just bite the bullet and fix #34 rather than having to do this AND later do #34.

**Comment by adamjanicki2** [comment]
> I don’t have a ton of time to review right now, but at first glance that seems …excessive. Maybe we should just bite the bullet and fix #34 rather than having to do this AND later do #34.

Sure sounds good to me, I'll take the conversation back over to #34 so we can land on a solution

**Pull request closed without merging**